### PR TITLE
Expanding the patterns

### PR DESCRIPTION
**Pull request by jeffrey-opdam**
- Patterns are being expanded into the teamproject
Which means a copy of the pattern is made, and {teamProject.Name} will be replaced.

### CODE REVIEW

File: src/drift.core/Extensions/StringExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+// </copyright>
+// -----------------------------------------------------------------------
+
+namespace Rangers.Antidrift.Drift.Core
+{    
+    using System.Linq.Dynamic.Core;
+    using System.Linq.Expressions;
+    using System.Text.RegularExpressions;
+
+    public static class StringExtensions
```
Comment (almguru) [inline src/drift.core/Extensions/StringExtensions.cs:17]: Public class is missing XML documentation.

File: src/drift.core/Extensions/StringExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    {
+        public static string Expand(this string target, TeamProject teamProject)
```
Comment (almguru) [inline src/drift.core/Extensions/StringExtensions.cs:19]: Public method is missing XML documentation.

File: src/drift.core/model/ApplicationGroup.cs [hunk 3]
```diff
@@ -11,17 +11,28 @@
...
     using System.Collections.Generic;
+    using System.Linq;
 
     public class ApplicationGroup
     {
         public string Descriptor { get; set; }
 
         public bool IsSpecial { get; set; }
 
+        public string Name { get; set; }
```
Comment (almguru) [inline src/drift.core/model/ApplicationGroup.cs:22]: Missing XML documentation.

File: src/drift.core/model/ApplicationGroup.cs [hunk 4]
```diff
@@ -11,17 +11,28 @@
...
+
+        public IList<Namespace> Namespaces { get; set; } = new List<Namespace>();
```
Comment (almguru) [inline src/drift.core/model/ApplicationGroup.cs:24]: Missing XML documentation.

File: src/drift.core/model/ApplicationGroup.cs [hunk 5]
```diff
@@ -11,17 +11,28 @@
...
+
```
Comment (almguru) [inline src/drift.core/model/ApplicationGroup.cs:25]: Missing XML documentation.

File: src/drift.core/model/Namespace.cs [hunk 6]
```diff
@@ -17,5 +19,15 @@ public class Namespace
         public string[] Deny { get; set; }
 
         public string Name { get; set; }
+
+        public Namespace Expand(TeamProject teamProject)
```
Comment (almguru) [inline src/drift.core/model/Namespace.cs:23]: Missing XML documentation.

File: src/drift.core/model/Namespace.cs [hunk 7]
```diff
@@ -17,5 +19,15 @@ public class Namespace
...
+        {
+            var result = new Namespace();
```
Comment (almguru) [inline src/drift.core/model/Namespace.cs:25]: Can be simplified to:
            return new Namespace
            {
                Allow = this.Allow.Select(a => a.Expand(teamProject)).ToArray(),
                Deny = this.Deny.Select(d => d.Expand(teamProject)).ToArray(),
                Name = this.Name.Expand(teamProject),
            };

File: src/drift.core/model/ApplicationGroup.cs [hunk 8]
```diff
@@ -11,17 +11,28 @@
...
         public string[] Members { get; set; } = System.Array.Empty<string>();
 
-        public string Name { get; set; }
+        public ApplicationGroup Expand(TeamProject teamProject)
+        {
+            var result = new ApplicationGroup();
```
Comment (almguru) [inline src/drift.core/model/ApplicationGroup.cs:30]: Can be simplified to:

            return new ApplicationGroup
            {
                Name = this.Name.Expand(teamProject),
                Namespaces = this.Namespaces.Select(n => n.Expand(teamProject)).ToList(),
                Members = this.Members.Select(m => m.Expand(teamProject)).ToArray(),
            };

File: src/drift.core/model/ApplicationGroup.cs [hunk 9]
```diff
@@ -11,17 +11,28 @@
...
         public bool IsSpecial { get; set; }
 
+        public string Name { get; set; }
+
+        public IList<Namespace> Namespaces { get; set; } = new List<Namespace>();
+
         public string[] Members { get; set; } = System.Array.Empty<string>();
 
-        public string Name { get; set; }
+        public ApplicationGroup Expand(TeamProject teamProject)
```
Comment (almguru) [inline src/drift.core/model/ApplicationGroup.cs:28]: Missing XML documentation

File: src/drift.core/model/Organization.cs [hunk 10]
```diff
@@ -25,6 +25,24 @@ public class Organization
 
         public IList<Team> Teams { get; } = new List<Team>();
 
+        public void Expand()
```
Comment (almguru) [inline src/drift.core/model/Organization.cs:28]: Missing XML documentation

File: src/drift.core/model/Pattern.cs [hunk 11]
```diff
@@ -21,5 +22,16 @@ public async virtual Task<IEnumerable<Deviation>> CollectDeviations(TeamProject
         {
             return await Task.FromResult(new List<Deviation>()).ConfigureAwait(false);
         }
+
+        public virtual Pattern Expand(TeamProject teamProject)
```
Comment (almguru) [inline src/drift.core/model/Pattern.cs:26]: Missing XML documentation

File: src/drift.core/model/Pattern.cs [hunk 12]
```diff
@@ -21,5 +22,16 @@ public async virtual Task<IEnumerable<Deviation>> CollectDeviations(TeamProject
...
+        {
+            return this.Expand(new Pattern(), teamProject);
+        }
+
+        public virtual Pattern Expand(Pattern pattern, TeamProject teamProject)
```
Comment (almguru) [inline src/drift.core/model/Pattern.cs:31]: Missing XML documentation

File: src/drift.core/model/SecurityPattern.cs [hunk 13]
```diff
@@ -74,5 +74,14 @@ public async override Task<IEnumerable<Deviation>> CollectDeviations(TeamProject
 
             return results;
         }
+    
+        public override Pattern Expand(TeamProject teamProject)
```
Comment (almguru) [inline src/drift.core/model/SecurityPattern.cs:78]: Missing XML documentation

**Review by almguru — APPROVED** [review]
I don't think my comments are critical, so we're good to go. At the same time ignoring XML doc comments would accumulate technical debts. I would tighten up build policy and disallow warning increase.

**Comment by jeffrey-opdam** [comment]
Currently prioritizing for getting the MVP done. So we can gather feedback.

**Pull request merged**